### PR TITLE
Check that source_bank datatype matches when reading from file

### PR DESCRIPTION
**Pull request by paulromano**
Some users have been running into segfaults when trying to create a source file that is then read by OpenMC because their file did not have all the proper source site attributes (compound datatype members in the HDF5 file). Because our source file format has changed a few times recently, this isn't too surprising. This PR adds a check to ensure that the datatype members in the source file match what is expected. If they don't match, an informative error message is given.

### CODE REVIEW

**Comment by gridley** [comment]
Similar to how the state point files have different versions, do you think it would be good to add a version to this file? I also like the approach of just checking if the right members are there, then working as expected if they're present. If it walks like a source file and it quacks like a source file, then it must be a source file. No need to version.

BTW, for reviewing functionality like this that's not regression tested, is it expected for reviewers to pull the code and test it? I don't have any source files of different versions lying around and can't test the new functionality.

**Comment by paulromano** [comment]
Yeah, that's a really good point @gridley. The nice thing about checking the attributes is that it tells you exactly what's missing, whereas if we were to do a version comparison, the error message would just be "version doesn't match" which isn't super helpful. So, I guess I'd vote to keep what's currently in this PR.

Good point on testing. It's your right as a reviewer to tell me "write a test!" Sorry for the laziness on my part -- I'll put something together.

**Comment by paulromano** [comment]
Ok, just added a test for this

File: tests/unit_tests/test_source_file.py [hunk 1]
```diff
@@ -43,3 +44,35 @@ def test_source_file(run_in_tmpdir):
     assert np.all(arr['wgt'] == 1.0)
     assert np.all(arr['delayed_group'] == 0)
     assert np.all(arr['particle'] == 0)
+
+
+def test_wrong_source_attributes(run_in_tmpdir):
+    # Create a source file with animal attributes
+    source_dtype = np.dtype([
+        ('platypus', '<f8'),
```
Comment (gridley) [inline tests/unit_tests/test_source_file.py:52]: Excellent examples :0

**Review by gridley — APPROVED** [review]
Looks good to me! I'll go ahead and merge.

**Pull request merged**